File: 0735-asteroid-collision/0735-asteroid-collision.cpp
class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        stack<int> st;
        for (auto e : asteroids) {
            if (e > 0) st.push(e);
            else {
                if (st.empty()) st.push(e);
                else {
                    bool isAsteroidDestroyed = false;
                    while (!st.empty() and (st.top() > 0)) {
                        if (abs(st.top()) == abs(e)) {
                            st.pop();
                            isAsteroidDestroyed = true;
                            break;
                        } else if (abs(st.top()) < abs(e)) {
                            st.pop();
                        } else {
                            isAsteroidDestroyed = true;
                            break;
                        }
                    }
                    if (!isAsteroidDestroyed) st.push(e);
                }
            }
        }
        vector<int> ans;
        while (!st.empty()) {
            ans.push_back(st.top());
            st.pop();
        }
        reverse(ans.begin(), ans.end());
        return ans;
    }
};

File: 0735-asteroid-collision/0735-asteroid-collision.java
class Solution {
    public int[] asteroidCollision(int[] asteroids) {
        Stack<Integer> stack = new Stack<>();
        for (int asteroid : asteroids) {
            if (asteroid > 0) {
                stack.push(asteroid);
            } else {
                if (stack.isEmpty() || stack.peek() < 0) {
                    stack.push(asteroid);
                } else {
                    boolean isExploded = false;
                    while (!stack.isEmpty() && stack.peek() > 0) {
                        int collideWith = stack.pop();
                        if (Math.abs(collideWith) == Math.abs(asteroid)) {
                            isExploded = true;
                            break;
                        } else if (Math.abs(asteroid) < Math.abs(collideWith)) {
                            isExploded = true;
                            stack.push(collideWith);
                            break;
                        } else {
                            continue;
                        }
                    }
                    if (!isExploded) stack.push(asteroid);
                }
            }
        }
        int[] ans = new int[stack.size()];
        int index = ans.length - 1;
        while (index >= 0) {
            ans[index] = stack.pop();
            --index;
        }
        return ans;
    }
}